### ISSUE TITLE
Crash on IoT secure tunnel on armv7l

### ISSUE THREAD

**Issue by fstehle** [awslabs/aws-iot-device-client#395]
**Describe the bug**

We run a cross-compiled version (on `armv7l`) of the AWS IoT device client.
Unfortunately the client crashes when we try to open an IoT secure tunnel:
```
...
2023-03-20T15:14:56.177Z [INFO]  {SecureTunnelingFeature.cpp}: Running Secure Tunneling!
2023-03-20T15:14:56.177Z [INFO]  {Main.cpp}: Client base has been notified that Secure Tunneling has started
2023-03-20T15:14:56.226Z [DEBUG] {SecureTunnelingFeature.cpp}: Subscribed to tunnel notification topic
2023-03-20T15:14:59.665Z [DEBUG] {SecureTunnelingFeature.cpp}: Received MQTT Tunnel Notification
terminate called after throwing an instance of 'std::logic_error'
  what():  basic_string::_M_construct null not valid
Aborted (core dumped)
```

We tried to debug the issue with gdb but had no success so far.
Does anyone know what the issue is?

**To Reproduce**

Compile the AWS IoT device client.
Run it with the following configuration:
```
{
  "endpoint": "xxx-ats.iot.eu-central-1.amazonaws.com",
  "cert": "/data/aws-iot-device-client-certs/cert.pem",
  "key": "/data/aws-iot-device-client-certs/key.pem",
  "root-ca": "/data/aws-iot-device-client-certs/root.ca.pem",
  "thing-name": "mything",
  "logging": {
    "level": "DEBUG",
    "type": "STDOUT"
  },
  "jobs": {
    "enabled": false
  },
  "tunneling": {
    "enabled": true
  },
  "device-defender": {
    "enabled": false
  },
  "fleet-provisioning": {
    "enabled": false
  },
  "samples": {
    "pub-sub": {
      "enabled": false
    }
  },
  "config-shadow": {
    "enabled": false
  },
  "sample-shadow": {
    "enabled": false
  },
  "secure-element": {
    "enabled": false
  }
}
```

**Expected behavior**

IoT tunnel is created

**Actual behavior**

aws-iot-device-client crashes

**Logs**

```
$ cat /var/run/aws-iot-device-client.conf
{
  "endpoint": "xxx-ats.iot.eu-central-1.amazonaws.com",
  "cert": "/data/aws-iot-device-client-certs/cert.pem",
  "key": "/data/aws-iot-device-client-certs/key.pem",
  "root-ca": "/data/aws-iot-device-client-certs/root.ca.pem",
  "thing-name": "mything",
  "logging": {
    "level": "DEBUG",
    "type": "STDOUT"
  },
  "jobs": {
    "enabled": false
  },
  "tunneling": {
    "enabled": true
  },
  "device-defender": {
    "enabled": false
  },
  "fleet-provisioning": {
    "enabled": false
  },
  "samples": {
    "pub-sub": {
      "enabled": false
    }
  },
  "config-shadow": {
    "enabled": false
  },
  "sample-shadow": {
    "enabled": false
  },
  "secure-element": {
    "enabled": false
  }
}

$ /sbin/aws-iot-device-client --config-file /var/run/aws-iot-device-client.conf
2023-03-20T15:14:55.798Z [INFO]  {Config.cpp}: Successfully fetched JSON config file: {
  "endpoint": "a3uxhu121qydvb-ats.iot.eu-central-1.amazonaws.com",
  "cert": "/data/aws-iot-device-client-certs/cert.pem",
  "key": "/data/aws-iot-device-client-certs/key.pem",
  "root-ca": "/data/aws-iot-device-client-certs/root.ca.pem",
  "thing-name": "mything",
  "logging": {
    "level": "DEBUG",
    "type": "STDOUT"
  },
  "jobs": {
    "enabled": false
  },
  "tunneling": {
    "enabled": true
  },
"device-defender": {
    "enabled": false
  },
  "fleet-provisioning": {
    "enabled": false
  },
  "samples": {
    "pub-sub": {
      "enabled": false
    }
  },
  "config-shadow": {
    "enabled": false
  },
  "sample-shadow": {
    "enabled": false
  },
  "secure-element": {
    "enabled": false
  }
}

2023-03-20T15:14:55.798Z [INFO]  {FileUtils.cpp}: Successfully create directory /home/root/.aws-iot-device-client/sample-shadow/ with required permissions 700
2023-03-20T15:14:55.798Z [INFO]  {Config.cpp}: ~/.aws-iot-device-client/sample-shadow/default-sample-shadow-document
2023-03-20T15:14:55.799Z [INFO]  {Config.cpp}: Succesfully create default file: /home/root/.aws-iot-device-client/sample-shadow/default-sample-shadow-document required for storage of shadow document
2023-03-20T15:14:55.799Z [DEBUG] {Config.cpp}: Did not find a runtime configuration file, assuming Fleet Provisioning has not run for this device
2023-03-20T15:14:55.799Z [DEBUG] {Config.cpp}: Did not find a http proxy config file /home/root/.aws-iot-device-client/http-proxy.conf, assuming HTTP proxy is disabled on this device
2023-03-20T15:14:55.799Z [DEBUG] {EnvUtils.cpp}: Updated PATH environment variable to: /usr/local/bin:/usr/bin:/bin:/usr/local/sbin:/usr/sbin:/sbin:/home/root/.aws-iot-device-client:/home/root/.aws-iot-device-client/jobs:/home/root:/home/root/jobs
2023-03-20T15:14:55.799Z [INFO]  {Main.cpp}: Now running AWS IoT Device Client version
2023-03-20T15:14:55.807Z [INFO]  {SharedCrtResourceManager.cpp}: SDK logging is disabled. Enable it with --enable-sdk-logging on the command line or logging::enable-sdk-logging in your configuration file
2023-03-20T15:14:55.808Z [DEBUG] {LockFile.cpp}: creating lockfile
2023-03-20T15:14:55.808Z [DEBUG] {Retry.cpp}: Retryable function starting, it will retry until success
2023-03-20T15:14:55.872Z [INFO]  {SharedCrtResourceManager.cpp}: Establishing MQTT connection with client id mything...
2023-03-20T15:14:56.176Z [INFO]  {SharedCrtResourceManager.cpp}: MQTT connection established with return code: 0
2023-03-20T15:14:56.176Z [INFO]  {SharedCrtResourceManager.cpp}: Shared MQTT connection is ready!
2023-03-20T15:14:56.177Z [INFO]  {Main.cpp}: Provisioning with Secure Elements is disabled
2023-03-20T15:14:56.177Z [INFO]  {Main.cpp}: Jobs is disabled
2023-03-20T15:14:56.177Z [INFO]  {Main.cpp}: Secure Tunneling is enabled
2023-03-20T15:14:56.177Z [INFO]  {Main.cpp}: Sensor Publish is disabled
2023-03-20T15:14:56.177Z [INFO]  {SharedCrtResourceManager.cpp}: Starting Device Client features.
2023-03-20T15:14:56.177Z [DEBUG] {FeatureRegistry.cpp}: Attempting to start Secure Tunneling
2023-03-20T15:14:56.177Z [INFO]  {SecureTunnelingFeature.cpp}: Running Secure Tunneling!
2023-03-20T15:14:56.177Z [INFO]  {Main.cpp}: Client base has been notified that Secure Tunneling has started
2023-03-20T15:14:56.226Z [DEBUG] {SecureTunnelingFeature.cpp}: Subscribed to tunnel notification topic
2023-03-20T15:14:59.665Z [DEBUG] {SecureTunnelingFeature.cpp}: Received MQTT Tunnel Notification
terminate called after throwing an instance of 'std::logic_error'
  what():  basic_string::_M_construct null not valid
Aborted (core dumped)
```

**Environment (please complete the following information):**
 - OS: Linux
 - Version Kernel 4.9.67
 - Architecture: armv7l
 - Device Client commit: [6633b3950182f1b1cd893c19d1dc1731b0eccd88](https://github.com/awslabs/aws-iot-device-client/commit/6633b3950182f1b1cd893c19d1dc1731b0eccd88)

**Additional context**

**Comment by fstehle** [comment]
Seems we are using a newer version of the `aws-iot-device-sdk-cpp-v2` in our build.
The `SecureTunnelBuilder` in https://github.com/aws/aws-iot-device-sdk-cpp-v2/blob/d498c9bbd74ed1672a71d0a1330a9789aaebcc14/secure_tunneling/include/aws/iotsecuretunneling/SecureTunnel.h needs to be used when using the newer version:
```
--- a/source/tunneling/SecureTunnelWrapper.cpp
+++ b/source/tunneling/SecureTunnelWrapper.cpp
@@ -23,21 +23,22 @@ SecureTunnelWrapper::SecureTunnelWrapper(
     const Aws::Iotsecuretunneling::OnStreamStart &onStreamStart,
     const Aws::Iotsecuretunneling::OnStreamReset &onStreamReset,
     const Aws::Iotsecuretunneling::OnSessionReset &onSessionReset)
-    : secureTunnel(new Aws::Iotsecuretunneling::SecureTunnel(
+    : secureTunnel(Aws::Iotsecuretunneling::SecureTunnelBuilder(
           allocator,
-          bootstrap,
+          *bootstrap,
           socketOptions,
           accessToken,
           localProxyMode,
-          endpoint,
-          rootCa,
-          onConnectionComplete,
-          onConnectionShutdown,
-          onSendDataComplete,
-          onDataReceive,
-          onStreamStart,
-          onStreamReset,
-          onSessionReset))
+          endpoint).
+          WithRootCa(rootCa).
+          WithOnConnectionComplete(onConnectionComplete).
+          WithOnConnectionShutdown(onConnectionShutdown).
+          WithOnSendDataComplete(onSendDataComplete).
+          WithOnDataReceive(onDataReceive).
+          WithOnStreamStart(onStreamStart).
+          WithOnStreamReset(onStreamReset).
+          WithOnSessionReset(onSessionReset).
+          Build())
```

However it still does not work with that version.
`OnSendDataComplete ` is mssing:
```
2023-03-21T09:16:14.018Z [INFO]  {SecureTunnelingFeature.cpp}: Running Secure Tunneling!
2023-03-21T09:16:14.019Z [INFO]  {Main.cpp}: Client base has been notified that Secure Tunneling has started
2023-03-21T09:16:14.081Z [DEBUG] {SecureTunnelingFeature.cpp}: Subscribed to tunnel notification topic
2023-03-21T09:16:34.195Z [DEBUG] {SecureTunnelingFeature.cpp}: Received MQTT Tunnel Notification
2023-03-21T09:16:34.195Z [DEBUG] {SecureTunnelingFeature.cpp}: Region=eu-central-1, Service=SSH
2023-03-21T09:16:34.510Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnConnectionComplete
2023-03-21T09:16:47.171Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnStreamStart
2023-03-21T09:16:47.189Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnDataReceive data.len=21
2023-03-21T09:16:47.189Z [DEBUG] {TcpForward.cpp}: Not connected yet. Saving the data to send
2023-03-21T09:16:47.190Z [DEBUG] {TcpForward.cpp}: TcpForward::OnConnectionResult
2023-03-21T09:16:47.190Z [DEBUG] {TcpForward.cpp}: Flushing send buffer
2023-03-21T09:16:47.286Z [DEBUG] {TcpForward.cpp}: TcpForward::OnReadable error_code=0
2023-03-21T09:16:47.286Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnTcpForwardDataReceive data.len=21
2023-03-21T09:16:47.291Z [DEBUG] {TcpForward.cpp}: TcpForward::OnReadable error_code=0
2023-03-21T09:16:47.291Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnTcpForwardDataReceive data.len=1048
2023-03-21T09:18:47.290Z [DEBUG] {TcpForward.cpp}: TcpForward::OnReadable error_code=1051
2023-03-21T09:18:47.290Z [DEBUG] {SecureTunnelingContext.cpp}: SecureTunnelingContext::OnTcpForwardDataReceive data.len=0
```

**Comment by gableshane** [comment]
Hi Fabian,

Sorry I didn't see this issue until just now! I will attempt to recreate the issue and see if I can find anything.

We have integration tests for Secure Tunneling in this repo but unfortunately they only run on amd64. cross-compiling the local proxy and device client times out the workflows.

Could you tell me which OS you are using so I can use the same for recreating the issue?

Thanks!
Shane

**Comment by fstehle** [comment]
Hey Shane,

thanks a lot.

We use the Yocto project for building the image system.
The recipes we use are from https://github.com/aws4embeddedlinux/meta-aws, eg. https://github.com/aws4embeddedlinux/meta-aws/blob/master/recipes-iot/aws-iot-device-client/aws-iot-device-client_1.8.bb

Best
Fabian

**Comment by aliraza1944** [comment]
Hi,

I am having the same issue, did you guys find a fix for that?

**Comment by aliraza1944** [comment]
Thanks for the update.

Few things that might help you reproduce the issue.

Build System: Ubuntu
End device: RPi4 (Arm v7)
recipe used: https://layers.openembedded.org/layerindex/recipe/304906/ (kirkstone branch)

**Comment by RogerZhongAWS** [comment]
For this issue: Due to a fairly recent update made to the CRT, the SecureTunnel constructor now takes in a client token parameter

```            
SecureTunnel(
Crt::Allocator *allocator,
Crt::Io::ClientBootstrap *clientBootstrap,
const Crt::Io::SocketOptions &socketOptions,
const std::string &accessToken,
const std::string &clientToken,
aws_secure_tunneling_local_proxy_mode localProxyMode,
const std::string &endpointHost,

const std::string &rootCa,
Crt::Http::HttpClientConnectionProxyOptions *httpClientConnectionProxyOptions,

OnConnectionSuccess onConnectionSuccess,
OnConnectionFailure onConnectionFailure,
OnConnectionComplete onConnectionComplete,
OnConnectionShutdown onConnectionShutdown,
OnSendDataComplete onSendDataComplete,
OnMessageReceived onMessageReceived,
OnDataReceive onDataReceive,
OnStreamStarted onStreamStarted,
OnStreamStart onStreamStart,
OnStreamStopped onStreamStopped,
OnStreamReset onStreamReset,
OnSessionReset onSessionReset,
OnStopped onStopped);
```

clientToken is passed as `nullptr` by the caller, which results in undefined behavior.

**Comment by RogerZhongAWS** [comment]
There is a second issue that uncovered itself after fixing the above.

There was some refactoring done to the tunnel `s_OnMessageReceived` callback function in the CRT. For backward compatibility, we construct an `aws_byte_buf` to pass back to the Device Client.

```
struct aws_byte_buf payload_buf;
AWS_ZERO_STRUCT(payload_buf);
payload_buf.allocator = NULL;
payload_buf.buffer = message->payload->ptr;
payload_buf.len = message->payload->len;
secureTunnel->m_OnDataReceive(payload_buf);
return;
```
There is also another validation function in the CRT that gets called when performing byte buf operations.
```
bool aws_byte_buf_is_valid(const struct aws_byte_buf *const buf) {
    return buf != NULL &&
           ((buf->capacity == 0 && buf->len == 0 && buf->buffer == NULL) ||
            (buf->capacity > 0 && buf->len <= buf->capacity && AWS_MEM_IS_WRITABLE(buf->buffer, buf->capacity)));
```

Because we never set the capacity, we observe that it is 0
```
(gdb) print /x *(aws_byte_buf*)0x7ffff7244b70
$8 = {len = 0x14, buffer = 0x7ffff49e404b, capacity = 0x0, allocator = 0x0}
```
and it fails the validation.

**Comment by RogerZhongAWS** [comment]
the fix has been released in the latest Device Client version (1.9)
